### PR TITLE
fix(client): Always generate package.json

### LINKED ISSUE #12296: Error on importing enum frontend  when Prisma client is generated on custom output
### Bug description

As already referenced in #1272 [here](https://github.com/prisma/prisma/issues/1272#issuecomment-760886780) by @joaopaulobdac there is an issue when importing an enum on frontends when the Prisma client is generated on a custom path.

To me, it's due to the fact that in the case of a custom output there is not the `package.json` file that would allow webpack (?) to get the appropriate file.

Prisma client generated in the default folder:
Capture d’écran de 2022-03-11 13-50-45

Prisma client generated in a custom folder (`output   = "../node_modules/.prisma-app1"`):
Capture d’écran de 2022-03-11 13-53-02

There is no `package.json` in the latter case.

If I copy the `package.json` file in the custom output, it works!

Content of the package.json:
```
{
  "name": ".prisma/client",
  "main": "index.js",
  "types": "index.d.ts",
  "browser": "index-browser.js"
}
```

After some search, it seems that this `package.json` is no more added to the Prisma client on purpose since release 2.11 to fix  #4048 (fixed by @timsuchanek).

Instead of removing `package.json`, could it be possible to rename the `name` property in the `package.json` file as [suggested by #4048 issuer](https://github.com/prisma/prisma/issues/4048#issue-730273330)?

### How to reproduce

Thanks to @joaopaulobdac

`./prisma/schema.prisma`
```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
  output   = "../generated"
}

model Post {
  id        Int      @id @default(autoincrement())
  title     String
  content   Letters?
  published Boolean  @default(false)
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?
}

model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
  posts Post[]
}

enum Letters {
  a
  b
  c
}
```

`./script.ts`
```ts
import { Letters } from './generated'

console.log(Letters)
```

Bundling
```console
$ yarn add esbuild

$ yarn esbuild --bundle script.ts
 > generated/index.js: error: Could not resolve "path" (set platform to "node" when building for node)
    20 │ const path = require('path')
       ╵                      ~~~~~~

 > generated/runtime/index.js: warning: Define "process.env.NODE_ENV" when bundling for the browser
    42005 │     if (process.env.NODE_ENV !== "test") {
          ╵         ~~~~~~~~~~~~~~~~~~~~

 > generated/runtime/index.js: error: Could not resolve "fs" (set platform to "node" when building for node)
    39 │   var fs3 = require("fs");
       ╵                     ~~~~

 > generated/runtime/index.js: error: Could not resolve "path" (set platform to "node" when building for node)
    160 │   var path3 = require("path");
        ╵                       ~~~~~~

 > generated/runtime/index.js: error: Could not resolve "child_process" (set platform to "node" when building for node)
    482 │   var cp = require("child_process");
        ╵                    ~~~~~~~~~~~~~~~

 > generated/runtime/index.js: error: Could not resolve "stream" (set platform to "node" when building for node)
    509 │   var stream = require("stream");
        ╵                        ~~~~~~~~

 > generated/runtime/index.js: error: Could not resolve "util" (set platform to "node" when building for node)
    510 │   var util2 = require("util");
        ╵                       ~~~~~~

 > generated/runtime/index.js: error: Could not resolve "os" (set platform to "node" when building for node)
    1703 │   var os2 = require("os");
         ╵                     ~~~~

 > generated/runtime/index.js: error: Could not resolve "tty" (set platform to "node" when building for node)
    1704 │   var tty = require("tty");
         ╵                     ~~~~~

 > generated/runtime/index.js: error: Could not resolve "crypto" (set platform to "node" when building for node)
    3556 │   var crypto_1 = __importDefault2(require("crypto"));
         ╵                                           ~~~~~~~~

 > generated/runtime/index.js: error: Could not resolve "assert" (set platform to "node" when building for node)
    8258 │     var assert = require("assert");
         ╵                          ~~~~~~~~
```

_Originally posted by @joaopaulobdac in https://github.com/prisma/prisma/issues/1272#issuecomment-760886780_

### Expected behavior

Importing an enum in a frontend application should not cause errors.

### Prisma information

```
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
  output   = "../generated"
}

model Post {
  id        Int      @id @default(autoincrement())
  title     String
  content   Letters?
  published Boolean  @default(false)
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?
}

model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
  posts Post[]
}

enum Letters {
  a
  b
  c
}
```

### Environment & setup

- OS: Linux Ubuntu
- Database: MySql
- Node.js version: v16.13.0

### Prisma Version

```
prisma                  : 3.10.0
@prisma/client          : 3.10.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine 73e60b76d394f8d37d8ebd1f8918c79029f0db86 (at node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli 73e60b76d394f8d37d8ebd1f8918c79029f0db86 (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine    : introspection-core 73e60b76d394f8d37d8ebd1f8918c79029f0db86 (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary           : prisma-fmt 73e60b76d394f8d37d8ebd1f8918c79029f0db86 (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash    : 73e60b76d394f8d37d8ebd1f8918c79029f0db86
Studio                  : 0.458.0
```

### PR DESCRIPTION
**Pull request by SevInf**
Fix #12296

### CODE REVIEW

File: packages/client/src/generation/generateClient.ts [hunk 1]
```diff
@@ -287,20 +287,20 @@ export async function generateClient({
...
-        main: 'index.js',
-        types: 'index.d.ts',
-        browser: 'index-browser.js',
-      },
-      null,
-      2,
-    )
-    await writeFile(packageJsonTargetPath, pkgJson)
-  }
+  // if (!generator?.isCustomOutput) {
```
Comment (danstarns) [inline packages/client/src/generation/generateClient.ts:290]: Are these comments left here for a reason?

File: packages/client/src/generation/generateClient.ts [hunk 2]
```diff
@@ -405,3 +405,16 @@ function validateDmmfAgainstDenylists(prismaClientDmmf: PrismaClientDMMF.Documen
 
   return errorArray.length > 0 ? errorArray : null
 }
+
+export function getPackageJsonName(outputPath: string): string {
```
Comment (danstarns) [inline packages/client/src/generation/generateClient.ts:409]: For me, `PackageJsonName` implies one of: 

```
"name": "@prisma/sdk",
```

```
"name": "@prisma/migrate",
```

```
"name": "@prisma/client",
```

```
"name": "xyz",
```

Is this function getting that `.name` value inside the package json?

File: packages/client/src/generation/generateClient.ts [hunk 3]
```diff
@@ -287,20 +287,18 @@ export async function generateClient({
...
-      },
-      null,
-      2,
-    )
-    await writeFile(packageJsonTargetPath, pkgJson)
-  }
+  const packageJsonTargetPath = path.join(finalOutputDir, 'package.json')
+  const pkgJson = JSON.stringify(
+    {
+      name: buildGeneratedPackageJsonName(finalOutputDir),
```
Comment (millsp) [inline packages/client/src/generation/generateClient.ts:293]: Potentially stupid question: Does the name of the package matter?

Comment (SevInf) [reply packages/client/src/generation/generateClient.ts:293]: I don't think it does, actually, unless users want to publish generated package, in which case they probably want to have their own name anyway. After thinking about it, we could probably just always leave `.prisma/client` there or even remove it completely.

**Comment by Jolg42** [comment]
@SevInf Check this 😉 https://github.com/prisma/prisma/blob/main/TESTING.md#ci---continuous-integration
If something is unclear, feel free to ping me / improve it

**Comment by SevInf** [comment]
Ecosystem test added in prisma/ecosystem-tests#2779

**Review by aqrln — APPROVED** [review]

**Pull request merged**